File: ios-native/Sources/FeatureTasks/TasksViewModel.swift
import CoreData
import Foundation
import Observation

public struct QuickAddState: Sendable, Equatable {
    public var visible = false
    public var text = ""
    public var busy = false

    public init() {}
}

@MainActor
@Observable
public final class TasksViewModel {

    private let taskRepository: TaskRepository
    private let navigationRepository: NavigationRepository
    private let authRepository: AuthRepository

    @ObservationIgnored private var started = false

    /// Ces trois-là ne sont pas recopiés dans un état local : ils sont lus
    /// directement dans les dépôts, qui sont eux-mêmes observables. Un miroir
    /// ici serait une seconde source de vérité à tenir synchronisée, ce que le
    /// `StateFlow` exposé tel quel du côté Android évite déjà.
    public var tasks: TasksState { taskRepository.state }
    public var navigation: NavigationState { navigationRepository.state }
    public var view: TaskView { taskRepository.view }

    public var quickAdd = QuickAddState()

    public init(
        taskRepository: TaskRepository,
        navigationRepository: NavigationRepository,
        authRepository: AuthRepository
    ) {
        self.taskRepository = taskRepository
        self.navigationRepository = navigationRepository
        self.authRepository = authRepository
    }

    public func start() {
        guard !started else { return }
        started = true
        Task { await taskRepository.refresh() }
        Task { await navigationRepository.refresh() }
    }

    /// Sélection dans le tiroir : le dépôt recharge la liste pour cette vue.
    public func select(_ view: TaskView) {
        taskRepository.select(view)
    }

    public func refresh() {
        Task { await taskRepository.refresh() }
    }

    public func setDone(_ id: String) {
        Task { _ = await taskRepository.setDone(id, done: true) }
    }

    public func showQuickAdd() {
        quickAdd = QuickAddState()
        quickAdd.visible = true
    }

    public func hideQuickAdd() {
        quickAdd = QuickAddState()
    }

    public func submitQuickAdd() {
        let text = quickAdd.text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty else { return }
        quickAdd.busy = true
        Task {
            _ = await taskRepository.quickAdd(text)
            quickAdd = QuickAddState()
        }
    }

    public func signOut() {
        Task { await authRepository.logout() }
    }
}

File: android-native/feature/tasks/src/main/kotlin/today/mindlog/todo/feature/tasks/TasksViewModel.kt
package today.mindlog.todo.feature.tasks

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import today.mindlog.todo.core.data.AuthRepository
import today.mindlog.todo.core.data.NavigationRepository
import today.mindlog.todo.core.data.NavigationState
import today.mindlog.todo.core.data.TaskRepository
import today.mindlog.todo.core.data.TaskView
import today.mindlog.todo.core.data.TasksState
import javax.inject.Inject

data class QuickAddState(
    val visible: Boolean = false,
    val text: String = "",
    val busy: Boolean = false,
)

@HiltViewModel
class TasksViewModel @Inject constructor(
    private val taskRepository: TaskRepository,
    private val navigationRepository: NavigationRepository,
    private val authRepository: AuthRepository,
) : ViewModel() {

    val tasks: StateFlow<TasksState> = taskRepository.state

    /** Contenu du tiroir, et vue sélectionnée : les deux viennent des dépôts. */
    val navigation: StateFlow<NavigationState> = navigationRepository.state
    val view: StateFlow<TaskView> = taskRepository.view

    private val _quickAdd = MutableStateFlow(QuickAddState())
    val quickAdd: StateFlow<QuickAddState> = _quickAdd.asStateFlow()

    init {
        viewModelScope.launch { taskRepository.refresh() }
        viewModelScope.launch { navigationRepository.refresh() }
    }

    /** Sélection dans le tiroir : le dépôt recharge la liste pour cette vue. */
    fun select(view: TaskView) = taskRepository.select(view)

    fun refresh() = viewModelScope.launch { taskRepository.refresh() }

    /** Page suivante ; le dépôt ignore l'appel si une page est déjà en vol. */
    fun loadMore() = taskRepository.loadMore()

    fun setDone(id: String) = viewModelScope.launch { taskRepository.setDone(id, done = true) }

    fun showQuickAdd() = _quickAdd.update { QuickAddState(visible = true) }
    fun hideQuickAdd() = _quickAdd.update { QuickAddState() }
    fun onQuickAddTextChange(value: String) = _quickAdd.update { it.copy(text = value) }

    fun submitQuickAdd() {
        val text = _quickAdd.value.text.trim()
        if (text.isEmpty()) return
        _quickAdd.update { it.copy(busy = true) }
        viewModelScope.launch {
            taskRepository.quickAdd(text)
            _quickAdd.value = QuickAddState()
        }
    }

    fun signOut() = viewModelScope.launch { authRepository.logout() }
}
